v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,238 @@
+// events are dropped if it exceeds the upload size limit or there is a marshalling problem.
+func (b *eventBuffer) compress(p *Plugin) {
+	// previous payload hasn't been sent and needs to be retried
+	if b.payload.bytesWritten != 0 {
+		return
+	}
+
+	eventLen := len(b.buffer)
+	if eventLen == 0 {
+		return
+	}
+
+	// Uploads at most the capacity of the buffer (buffer_size_limit_events)
+	for range eventLen {
+		select {
+		case event := <-b.buffer:
+			// an event could already have been serialized if an upload reached the limit or failed
+			if event.serialized == nil {
+				var err error
+				event.serialized, err = json.Marshal(event.EventV1)
+				if err != nil {
+					p.incrMetric(logEncodingFailureCounterName)
+					p.logger.Error("encoding failure: %v, dropping event with decision ID: %v.", err, event.EventV1.DecisionID)
+					continue
+				}
+			}
+
+			if int64(len(event.serialized)) >= b.uploadSizeLimitBytes {
+				if event.NDBuiltinCache == nil {
+					p.logger.Error("upload event size (%d) exceeds upload_size_limit_bytes (%d), dropping event with decision ID: %v.",
+						int64(len(event.serialized)), b.uploadSizeLimitBytes, event.EventV1.DecisionID)
+					continue
+				}
+
+				// Attempt to drop the ND cache to reduce size. If it is still too large, drop the event.
+				event.NDBuiltinCache = nil
+				var err error
+				event.serialized, err = json.Marshal(event.EventV1)
+				if err != nil {
+					p.incrMetric(logEncodingFailureCounterName)
+					p.logger.Error("encoding failure: %v, dropping event with decision ID: %v.", err, event.EventV1.DecisionID)
+					continue
+				}
+				if int64(len(event.serialized)) > b.uploadSizeLimitBytes {
+					p.logger.Error("upload event size (%d) exceeds upload_size_limit_bytes (%d), dropping event with decision ID: %v.",
+						int64(len(event.serialized)), b.uploadSizeLimitBytes, event.EventV1.DecisionID)
+					continue
+				}
+
+				p.incrMetric(logNDBDropCounterName)
+				p.logger.Error("ND builtins cache dropped from this event to fit under maximum upload size limits. Increase upload size limit or change usage of non-deterministic builtins.")
+			}
+
+			if !b.payload.capacity(len(event.serialized), b.uploadSizeLimitBytes) {
+				// add the event that doesn't fit back into the buffer
+				b.Push(p, event)
+				break
+			}
+
+			if err := b.payload.add(event); err != nil {
+				// corrupted payload, record failure and reset
+				b.payload.reset()
+				p.incrMetric(logEncodingFailureCounterName)
+				p.logger.Error("encoding failure: %v, dropping event with decision ID: %v.", err, event.EventV1.DecisionID)
+				continue
+			}
+		default:
+			break
+		}
+	}
+
+	if err := b.payload.close(); err != nil {
+		// corrupted payload, record failure and reset
+		b.payload.reset()
+		p.incrMetric(logEncodingFailureCounterName)
+		p.logger.Error("encoding failure: %v, dropping multiple events.")
+		return
+	}
+}
+
+// payload contains the compressed JSON array to be uploaded
+type payload struct {
```

Comment:
Is there a reason why we're not reusing the `chunkEncoder` instead?

Reply:
If there was a reason, it isn't there anymore haha Good idea, updated to use the `chunkEncoder` instead.